base/compiler/abstractinterpretation.jl

```diff
@@ -2681,12 +2681,15 @@ end
 # refine the result of instantiation of partially-known type `t` if some invariant can be assumed
 function refine_partial_type(@nospecialize t)
     t′ = unwrap_unionall(t)
-    if isa(t′, DataType) && t′.name === _NAMEDTUPLE_NAME && length(t′.parameters) == 2 &&
-        (t′.parameters[1] === () || t′.parameters[2] === Tuple{})
+    if isa(t′, DataType) && t′.name === _NAMEDTUPLE_NAME && length(t′.parameters) == 2
         # if the first/second parameter of `NamedTuple` is known to be empty,
         # the second/first argument should also be empty tuple type,
         # so refine it here
-        return Const(NamedTuple())
+        if t′.parameters[2] === Tuple{}
+            return Const(NamedTuple{(), Tuple{}}(()))
+        elseif t′.parameters[1] === ()
+            return Union{NamedTuple{(), Union{}}, NamedTuple{(), Tuple{}}}
```

Comment:
I think we can make `NamedTuple{(), Union{}}` still an error, since `NamedTuple{(),Tuple{Union{}}}` etc. was an error before.